### PR TITLE
Remove importance of order of rule in RestoreSession

### PR DESCRIPTION
**Pull request by hossainemruz**
Tasks:
- [x] Remove importance of rule order
- [x] Add validator to ensure followings,
	- [x] There is at most one rule with empty targetHosts field.
	- [x] No two rule with non-empty targetHosts matches for a host.
	- [x] If snapshot is specified in a rule then paths is not specified.
- [x] Update concept doc to make rules behavior clear

Fixes: https://github.com/stashed/stash/issues/790

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -17,19 +22,22 @@ require (
 	github.com/gopherjs/gopherjs v0.0.0-20190430165422-3e4dfb77656c // indirect
 	github.com/graymeta/stow v0.0.0-00010101000000-000000000000
 	github.com/json-iterator/go v1.1.6
+	github.com/mitchellh/go-homedir v1.1.0 // indirect
 	github.com/onsi/ginkgo v1.8.0
 	github.com/onsi/gomega v1.5.0
 	github.com/pkg/errors v0.8.1
 	github.com/prometheus/client_golang v0.9.3-0.20190127221311-3c4408c8b829
 	github.com/prometheus/common v0.4.0 // indirect
+	github.com/sgotti/glide-vc v0.1.0 // indirect
```
Comment (tamalsaha) [inline go.mod]: Why do we have a dependency on this? :open_mouth:

File: go.mod [hunk 2]
```diff
@@ -4,11 +4,16 @@ go 1.12
 
 require (
 	github.com/Azure/go-autorest v11.1.2+incompatible // indirect
+	github.com/Masterminds/glide v0.13.2 // indirect
```
Comment (tamalsaha) [inline go.mod]: Why do we have a dependency on this? :open_mouth:

File: go.mod (hunk 1 above)
Comment (hossainemruz) [reply go.mod]: got removed after running `go mod tidy`

File: go.mod (hunk 2 above)
Comment (hossainemruz) [reply go.mod]: go removed after running `go mod tidy`

File: apis/stash/v1beta1/validator.go [hunk 3]
```diff
@@ -1,11 +1,69 @@
...
 
+import "fmt"
+
 // TODO: complete
 func (r BackupSession) IsValid() error {
 	return nil
 }
 
 // TODO: complete
 func (r RestoreSession) IsValid() error {
```
Comment (tamalsaha) [inline apis/stash/v1beta1/validator.go]: We need tests for this.
https://dave.cheney.net/2013/06/09/writing-table-driven-tests-in-go

**Pull request merged**